File: matlab/BuildFactorCache.m
%% header
% build factor score cache for all factors
% Date: 7/3/2015

%%
addpath('D:/ChinaA/source' ,'D:/ChinaA/source/Common', 'D:/ChinaA/source/FactorLib')
%initEnv

A50PPCache = 'd:\ChinaA\data\Universes\A50PP';
%a50PP = PortfolioProviders.GetA50();
%save(A50PPCache, 'a50PP');
%load(A50PPCache);  % load a50PP from cache

zz800PPCache = 'd:\ChinaA\data\Universes\zz800PP';
%zz800PP = PortfolioProviders.GetZhongZheng800();
%save(zz800PPCache, 'zz800PP');
load(zz800PPCache);  % load pp from cache
%pf = GetPortfolioAsofFixed(zz800PP, '20141231');
% 

WkDts = QDate.WeekEndsBtw('20090401', '20141231');
MonDts = QDate.MonthEndsBtw('20090401', '20141231');
CacheDts = QDate.UnionDistinct(WkDts,MonDts);

BP = Factor('BP','Book/Price',@BPCalc);
zBP=Z(BP, false, zz800PP);
zBP_SN=Z(BP, true, zz800PP);

% a50PP 2009-04-01 ~ 2014-12-31
%CalcScoresAndSave(BP,a50PP.AvailableDates,a50PP); % tests
CalcScoresAndSave(BP,CacheDts,zz800PP); % tests
CalcScoresAndSave(zBP,CacheDts,zz800PP); % tests
CalcScoresAndSave(zBP_SN,CacheDts,zz800PP); % tests

File: buildFactorCache.py
""" 
Build factor score cache for all factors 
"""

import os
import sys
from datetime import datetime
import cPickle as pickle
import common.GlobalConstant as GlobalConstant
from common.Factor import Factor
from common.PortfolioProviders import PortfolioProviders
import common.QDate as QDate
from common.Factor import Factor
import common.Factorlib.BPCalc as BPCalc



def buildFactorCache():
    """ Build factor score cache for all factors """

    # Avoid error `maximum recursion depth exceeded`
    sys.setrecursionlimit(1000)

    # Retrieve all PortfolioProviders
    a50PP, zz800PP = getAllPP()



def getAllPP():
    """ Retrieve all PortfolioProviders 
        Hits flatfile cache first, before querying database
    """
    # Construct path to cache file
    universesPath = os.path.join(GlobalConstant.DATA_DIR, 'Universes')
    a50ppCachePath = os.path.join(universesPath, 'a50PP.dat')
    zz800ppCachePath = os.path.join(universesPath, 'zz800PP.dat')

    # Load cache if cache file exist
    if os.path.isfile(a50ppCachePath) & os.path.isfile(zz800ppCachePath):
        a50PP, zz800PP = getAllPPFromCache()
    # Build cache if not exist
    else:
        a50PP, zz800PP = getAllPPFromDb()

    return (a50PP, zz800PP)


def getAllPPFromCache():
    """ Retrieve all PortfolioProviders from flat file cache """
    # Construct path to cache file
    universesPath = os.path.join(GlobalConstant.DATA_DIR, 'Universes')
    a50ppCachePath = os.path.join(universesPath, 'a50PP.dat')
    zz800ppCachePath = os.path.join(universesPath, 'zz800PP.dat')

    # Decode pickled objects
    f = file(a50ppCachePath, 'rb')
    a50PP = pickle.load(f)
    f = file(zz800ppCachePath, 'rb')
    zz800PP = pickle.load(f)

    return (a50PP, zz800PP)


def getAllPPFromDb():
    """ Retrieve all PortfolioProviders from Databases """
    
    # Construct path to cache file
    universesPath = os.path.join(GlobalConstant.DATA_DIR, 'Universes')
    a50ppCachePath = os.path.join(universesPath, 'a50PP.dat')
    zz800ppCachePath = os.path.join(universesPath, 'zz800PP.dat')

    # Grab data
    a50PP = PortfolioProviders.getA50()
    zz800PP = PortfolioProviders.getZhongZheng800()

    # Write data to disk
    if not os.path.exists(universesPath):
        os.makedirs(universesPath)
    with open(a50ppCachePath, 'w') as fp:
        pickle.dump(a50PP, fp)
    with open(zz800ppCachePath, 'w') as fp:
        pickle.dump(zz800PP, fp)

    return (a50PP, zz800PP)


# # Build Factor cache
buildFactorCache()

# Setup datetimes
WeekDts = QDate.WeekEndsBtw( datetime.strptime('20090401', '%Y%m%d'), datetime.strptime('20141231', '%Y%m%d') )
MonthDts = QDate.MonthEndsBtw( datetime.strptime('20090401', '%Y%m%d'), datetime.strptime('20141231', '%Y%m%d') )
CacheDts = QDate.UnionDistinct(WeekDts, MonthDts)


#!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
# Bugs in imported modules, unable to continue
#!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!

# Setup PortfolioProviders
a50PP, zz800PP = getAllPP()
BP = Factor('BP', 'Book/Price', BPCalc, zz800PP)
zBP = BP.Z(False, zz800PP)
zBP_SN = BP.Z(True, zz800PP)

# Calculate Scores and Save
BP.CalcScoresAndSave(CacheDts, zz800PP)
zBP.CalcScoresAndSave(CacheDts, zz800PP)
zBP_SN.CalcScoresAndSave(CacheDts, zz800PP)